### PR TITLE
[Diffusion] Preserve per-sample rollout trajectories across multi-output merge

### PR DESCRIPTION
**Pull request by CjhHa1**
## Motivation

Reported in #34000 (Bug 1). **This is a silent numerical corruption bug** — no exception, no warning; per-sample rollout log-probs are simply wrong.

A `num_outputs_per_prompt=K` request runs as K per-output forwards, each producing its own `rollout_trajectory_data` from its own `x_T` slice and per-step SDE noise. Two sites collapse those K distinct trajectories into K copies of output 0's:

1. `GPUWorker._merge_expanded_singletons` keeps the trajectory of the **first** per-output batch only:

   ```python
   if (
       merged.rollout_trajectory_data is None
       and output_batch.rollout_trajectory_data is not None
   ):
       merged.rollout_trajectory_data = output_batch.rollout_trajectory_data
   ```

   Note the asymmetry: `output`, `trajectory_latents`, `noise_pred` and `trajectory_decoded` are all concatenated along the batch dim in `_finalize_expanded_parts`. Only `rollout_trajectory_data` is treated as a singleton.

2. `DiffGenerator._result_common` slices `samples` and `metrics` per `output_index` but hands the whole trajectory to **every** `GenerationResult`.

So every sample in the group carries output 0's `rollout_log_probs`. For GRPO-style post-training the per-sample log-probs become identical, the mean-zero advantages cancel, and the gradient vanishes — reward stays flat with `grad_norm ≈ 0`.

Both entry points into the merge are affected, including the sequential path added in #30683: `_forward_group` and `_collect_sequential_outputs` both call `_merge_expanded_output_batches`.

**This also fixes `rollout_api` (the actively maintained rollout interface) without touching it.** `_slice_rollout_trajectory_for_sample` already slices per-sample, but every slice is guarded on `shape[0] == batch_size`. Against the collapsed `[1, ...]` tensor that guard doesn't match, so `_extract_single_sample_tensor` returns the tensor **whole** and the slicing silently no-ops. That existing code already expects a `[K, ...]` trajectory — this PR makes the merge actually produce one.

## Modifications

- `gpu_worker.py`: collect `rollout_trajectory_data` into `_ExpandedOutputParts` and concatenate it along dim 0 in `_finalize_expanded_parts`, alongside the other per-output tensors. New module-level `_concat_rollout_trajectory_data` handles the nested structure; `timesteps` / `sigmas` / `denoising_env` describe the shared schedule and conditioning, so they come from the first output.
- `diffusion_generator.py`: `_result_common` narrows the grouped trajectory to the requested `output_index`, keep-dim, so a grouped result carries the same `[1, ...]` shape as an unexpanded one.

`output` dim 0 and the trajectory dim 0 are produced by the same per-output loop, so they agree by construction for every K — which is what makes `rollout_api`'s existing `shape[0] == batch_size` guard start matching.

**One deliberate behavior change:** when only *some* outputs in a group carry a trajectory, the merge now yields `None` rather than the single present row. Labeling one sample's trajectory as the whole group's is the same broadcast bug in a different disguise, so dropping it is preferable to misaligning it. Happy to change this if you'd prefer it preserved.

## Accuracy Test

Observed with K=3, T=4, per-output values tagged 1/2/3:

```
                          merged log_probs   result[0]  result[1]  result[2]
before                        (1, 4)             1          1          1     <- all output 0
after                         (3, 4)             1          2          3
```

`rollout_api` path, with `_slice_rollout_trajectory_for_sample` unchanged:

```
collapsed [1,...] merge (before)  -> per-sample values ['1','1','1']  ALL IDENTICAL
concatenated [K,...] merge (after) -> per-sample values ['1','2','3']  DISTINCT
```

Unchanged paths, checked before/after: K=1 trajectory shape `(1,4)`; non-rollout requests `None`; `output` tensor merge `(3,3)`.

**Verification caveat, please read:** I could not run the pytest suite against `main` — the sglang install I have is `0.5.12.post1`, whose tree predates the current layout, and my only box is a CPU node where importing `sglang.multimodal_gen` off network storage takes >10 minutes. The numbers above come from extracting the real functions under test from both `main` and this branch via `git show` and executing them side by side against stubs. **The added unit test has not been executed under a real sglang install** — please treat it as unvalidated and let CI be the first real run.

## Checklist

- [x] Format the code with pre-commit hooks (isort / black / ruff clean on the changed files)
- [x] Add unit tests (`test_multi_output_rollout_trajectory.py`) — but see the verification caveat above
- [x] Linked the tracking issue (#34000)

---
### CI States

Latest PR Test (Base): :x: [Run #35328611696](https://github.com/sgl-project/sglang/actions/runs/35328611696)
Latest PR Test (Extra): :x: [Run #35328611442](https://github.com/sgl-project/sglang/actions/runs/35328611442)
Latest PR Test (AMD ROCm 10): :x: [Run #35328611580](https://github.com/sgl-project/sglang/actions/runs/35328611580)

### CODE REVIEW

**Comment by CjhHa1** [comment]
Rebased onto current `main`. The first-wins merge in `_merge_expanded_singletons` is still there, so this is still a real bug: `num_outputs_per_prompt=K` still collapses per-sample rollout trajectories to output 0, and `_slice_rollout_trajectory_for_sample` in `rollout_api` still no-ops because `shape[0] != batch_size`.

Conflict resolution was import-only (`_replace_sampling_params_for_prompt` kept; helpers added next to it).

**Comment by CjhHa1** [comment]
@mickqian @ping1jing2 @HaiShaw @yichiche @AgainstEntropy @BBuf this is rebased onto current `main` and still fixes a silent correctness bug: `num_outputs_per_prompt=K` still collapses per-sample rollout trajectories to output 0. That also breaks `rollout_api`, not just DiffGenerator. Happy to adjust if you'd rather land the merge half first.

File: python/sglang/multimodal_gen/runtime/entrypoints/diffusion_generator.py [hunk 1]
```diff
@@ -101,6 +108,68 @@ def _replace_sampling_params_for_prompt(
...
+
+    A ``num_outputs_per_prompt=K`` request merges its K per-output trajectories
+    into batch-dim-K tensors, but each ``GenerationResult`` describes a single
+    sample. Without this narrowing every result in the group reports the whole
+    group's trajectory, so per-sample log-probs are indistinguishable.
+
+    The row is kept as ``[1, ...]`` so a grouped result has the same shape as an
+    unexpanded one. ``timesteps`` / ``sigmas`` / ``denoising_env`` are
+    group-shared and pass through unchanged.
+    """
```
Comment (niehen6174) [inline python/sglang/multimodal_gen/runtime/entrypoints/diffusion_generator.py]: Helper docstrings are a bit long — the GRPO story is already on the issue. A one-liner each should be enough?

File: python/sglang/multimodal_gen/runtime/managers/gpu_worker.py [hunk 2]
```diff
@@ -118,6 +123,80 @@ class _ExpandedOutputParts:
...
+    ``[K, ...]`` batch, so it is dropped rather than silently misaligned.
+    """
+    if not tensors or any(tensor is None for tensor in tensors):
+        return None
+    return torch.cat(tensors, dim=0)
+
+
+def _concat_rollout_trajectory_data(
+    per_output: list[RolloutTrajectoryData | None],
+) -> RolloutTrajectoryData | None:
```
Comment (niehen6174) [inline python/sglang/multimodal_gen/runtime/managers/gpu_worker.py]: These three helpers only know RolloutTrajectoryData. Maybe they fit better on the dataclass (or next to rl_dataclasses.py), with a one-line call at the merge / _result_common? Not blocking.

File: python/sglang/multimodal_gen/runtime/entrypoints/diffusion_generator.py (hunk 1 above)
Comment (CjhHa1) [reply python/sglang/multimodal_gen/runtime/entrypoints/diffusion_generator.py]: Updated — the helper docstrings are now one line, and the detailed motivation remains in the issue/PR description.

File: python/sglang/multimodal_gen/runtime/managers/gpu_worker.py (hunk 2 above)
Comment (CjhHa1) [reply python/sglang/multimodal_gen/runtime/managers/gpu_worker.py]: Updated — moved the trajectory concat/select helpers next to the dataclasses in `rl_dataclasses.py`; the merge and result assembly now each have a one-line call.

**Comment by niehen6174** [comment]
could you resolve the conflict with main?
